lib/pure/options.nim

```diff
@@ -176,22 +297,240 @@ proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+  # runnableExample couldn't make sense of this macro..
+  var
+    noneCase: NimNode = nil
+    elseCase: tuple[ident, body: NimNode] = (ident: nil, body: nil)
+    cases: seq[tuple[ident, check, body: NimNode]] = @[]
+  for optionCase in body:
+    case optionCase.kind:
+    of nnkCall:
+      if $optionCase[0] != "none":
+        if $optionCase[0] != "some":
+          error "Only \"none\" and \"some\" are allowed as case labels",
+            optionCase[0]
+        else:
+          error "Only \"none\" is allowed to not have arguments", optionCase[0]
+      elif noneCase != nil:
+        error "Only one \"none\" case is allowed, " &
+          "previously defined \"none\" case at: " & lineInfo(noneCase),
+          optionCase[0]
+      else:
+        noneCase = optionCase[1]
+    of nnkCommand:
+      if $optionCase[0] != "some":
+        if $optionCase[0] != "none":
+          error "Only \"none\" and \"some\" are allowed as case labels",
+            optionCase[0]
+        else:
+          error "Only \"some\" is allowed to have arguments", optionCase[0]
+      else:
+        if optionCase[1].kind != nnkInfix:
+          if elseCase.body != nil:
+            error "Only one \"some\" case without a check is allowed, " &
+              "previously defined \"some\" case at: " & lineInfo(elseCase.body),
+              optionCase[0]
+          else:
+            if $optionCase[1] != "_":
+              elseCase = (ident: optionCase[1], body: optionCase[2])
+            else:
+              elseCase = (ident: nil, body: optionCase[2])
+        else:
+          if $optionCase[1][0] != "of":
+            error "Invalid separator, only \"of\" allowed", optionCase[1][0]
+          else:
+            if $optionCase[1][1] != "_":
+              cases.add (ident: optionCase[1][1], check: optionCase[1][2],
+                body: optionCase[2])
+            else:
+              cases.add (ident: nil, check: optionCase[1][2],
+                body: optionCase[2])
+    else:
+      error "Unrecognized structure of cases", optionCase
+  if noneCase == nil:
+    error "Must have a \"none\" case"
+  if elseCase.body == nil and cases.len == 0:
+    error "Must have a \"some\" case"
+  var
+    optionSym = genSym(nskLet)
+    value = nnkDotExpr.newTree(optionSym, newIdentNode("val"))
+  var caseStmt = nnkCaseStmt.newTree(
+    value
+  )
+  for c in cases:
+    var ofBranch = nnkOfBranch.newTree(
+      c.check,
+      newStmtList()
+    )
+    if c.ident != nil:
+      ofBranch[1].add newLetStmt(c.ident, value)
+    ofBranch[1].add c.body
+    caseStmt.add ofBranch
+  if elseCase.body != nil:
+    if elseCase.ident != nil:
+      caseStmt.add nnkElse.newTree(newStmtList(
+        newLetStmt(elseCase.ident, value),
+        elseCase.body
+      ))
+    else:
+      caseStmt.add nnkElse.newTree(elseCase.body)
+  result = quote do:
+    (proc (): auto =
+      let `optionSym` = `option`
+      if `optionSym`.isSome:
+        `caseStmt`
```

Comment:
This fails on values that can't be used in case statements.

Reply:
Hmm, that's true.. Maybe the `of` keyword here could signify a case statement is generated, but if you used something like `if` there instead it would generate an if-statement?